### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by shreyash-x**
Code Review

### CODE REVIEW

File: experiment/posttest.json [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+               
+            },
+            "difficulty": "intermediate"
+        },
+        {
+            "question": "3.What is the full form of PIPO",
+            "answers": {
+                "a": "Parallel In Parallel Out ",
+                "b": "Serial In Parllel Out ",
+                "c": "Parallel In Serial Out",
```
Comment (raj-vlabs) [inline experiment/posttest.json:35]: Is it possible to have better alternatives for options b and c, they don't even reduce to PIPO as an acronym

File: experiment/procedure.md [hunk 2]
```diff
@@ -1,7 +1,5 @@
-1. With The help of Theory Page , try to build the Register Circuit.
-2. First of all build the Simple Shift Register.
-3. Then try to build the Parallel Load shift Register.
-4. If You are unable to make it , load or export the circuit directly and use it.
-5. Use time variant input pulses to analyze the output.
-6. There is an import feature, which lets you construct building blocks. Click on import and then click on circuit board to see the imported block. For more details, refer import section in manual.
-
+1. Firstly learn everything from the 'Demo' section.
+2. Then, test your understanding about Latch and Flip Flops in the 'Pretest' section.
+3. Then in the 'Practice' section in the task-1 "Serial Register", make simple SISO Register using logic gates and D flip-flops and test it with time-varying clock input.
```
Comment (raj-vlabs) [inline experiment/procedure.md]: 1. 1 and 2 are not required here. The procedure exclusively deals with the demo and simulation part. Here it looks like the procedure is explaining how to use the experiment.
2. Please elaborate points 3 an 4 with explanations of what needs to be done, why that needs to be done and what needs to be observed

File: experiment/simulation/demo-parallel-register.html [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    <link rel="stylesheet" href="css/main.css">
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/gsap/3.9.1/gsap.min.js"></script>
+
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/gsap/3.9.1/MotionPathPlugin.min.js"></script>
+    <title>
+        Registers: Demo
+    </title>
+</head>
+
+<body onload="myFunction()">
```
Comment (raj-vlabs) [inline experiment/simulation/demo-parallel-register.html]: What is myFunction? What does it do and where is it defined?

File: experiment/simulation/demo-parallel-register.html [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            </div>
+        </div>
+
+        <div class="columns is-centered">
+            <div class="column">
+                <div class="v-instruction-container instructions-box">
+                    <div class="v-instruction-title" id="TaskTitle">PIPO shift register</div>
+                    <div class="v-instruction-text" id="TaskDescription">
+                        Simulation of a PIPO shift register built using D flip-flops.
+                        <p></p>
```
Comment (raj-vlabs) [inline experiment/simulation/demo-parallel-register.html]: Why are we using empty paragraphs here?

File: experiment/simulation/demo-parallel-register.html [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        </div>
+    </div>
+    
+    <div class="columns is-centered">
+        <div class="column is-one-quarter" style="margin:auto;text-align:center">
+
+        </div>
+        <div class="column" style="text-align:center">
+            <div class="v-select">
+                <select autocomplete="off" name="speed" onchange="SetSpeed(this.value)" id="speed">
```
Comment (raj-vlabs) [inline experiment/simulation/demo-parallel-register.html]: SetSpeed should be named setSpeed

File: experiment/simulation/demo-parallel-register.html [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                    <option value="1">1x</option>
+                    <option value="2">2x</option>
+                    <option value="4">4x</option>
+                </select>
+            </div>
+            <button class="v-button" id="play/pause" onclick="button()">Start</button>
```
Comment (raj-vlabs) [inline experiment/simulation/demo-parallel-register.html]: Please do not use a / in the identifier.
Why is the function named as button? What am I supposed to understand about this function from its name?

File: experiment/simulation/demo-parallel-register.html [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+            <button class="v-button" id="restart" onclick="restartCircuit()">Reset</button>
+        </div>
+        <div class="column is-one-third">
```
Comment (raj-vlabs) [inline experiment/simulation/demo-parallel-register.html]: Why do we have an empty column here?

File: experiment/simulation/demo-parallel-register.html [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+
+        </div>
+    </div>
+    <div class="columns is-centered is-flex is-flex-direction-row is-flex-wrap-wrap is-align-content-start">
+        <div class="column is-9-desktop is-12-tablet is-12-mobile">
+            <div id="circuit-board">
+                <div id="working-area">
+
+                    <svg class="svg" xmlns="http://www.w3.org/2000/svg" width="100%" height="120%"
```
Comment (raj-vlabs) [inline experiment/simulation/demo-parallel-register.html]: Just wondering if it's possible for this svg to be defined in a separate file and imported here. It's adding a lot of irrelevant details to this file.

File: experiment/simulation/demo-parallel-register.html [hunk 9]
```diff
@@ -0,0 +1,231 @@
+<!DOCTYPE html>
+<html>
+<html lang="en">
```
Comment (raj-vlabs) [inline experiment/simulation/demo-parallel-register.html:2]: Why do we have 2 html tags here?

File: experiment/simulation/demo-series-register.html [hunk 10]
```diff
@@ -0,0 +1,179 @@
+<!DOCTYPE html>
+<html>
+<html lang="en">
```
Comment (raj-vlabs) [inline experiment/simulation/demo-series-register.html]: This file has the same problems as the demo-parallel-register.html, please resolve those comments for both the files

File: experiment/simulation/demo-series-register.html [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+                        <p>
+							This is a SISO shift register.</p><p> It has a Serial Input(SI) bit. There is a clock which can be started by the user but it's cycle cannot be controlled.</p>
+                        <p>There is one output(SO) </p>
+					 </p>
+                    </div>
+                </div>
+            </div>
+        </div>
+    </div>
+    <br>
```
Comment (raj-vlabs) [inline experiment/simulation/demo-series-register.html]: Please do not use br tag

File: experiment/simulation/js/animation-parallel-register.js [hunk 12]
```diff
@@ -0,0 +1,807 @@
...
+let textI1 = document.createElementNS(svgns, "text");
+let textI2 = document.createElementNS(svgns, "text");
+let textI3 = document.createElementNS(svgns, "text");
+let textI4 = document.createElementNS(svgns, "text");
+let textClock = document.createElementNS(svgns, "text");
+let textO1 = document.createElementNS(svgns, "text");
+let textO2 = document.createElementNS(svgns, "text");
+let textO3 = document.createElementNS(svgns, "text");
+let textO4 = document.createElementNS(svgns, "text");
+let textSelect = document.createElementNS(svgns, "text");
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:32]: Please see how many of these 10 declarations be const

File: experiment/simulation/demo-parallel-register.html [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+                            d="M 120 20 L 120 240 L 250 240 " />
+                        <path id="path15" fill="none" stroke="currentColor" stroke-width="3"
+                            d="M 120 20 L 120 440 L 250 440 " />
+                        <path id="path16" fill="none" stroke="currentColor" stroke-width="3"
+                            d="M 120 20 L 120 640 L 250 640 " />
+                        <path id="path12" fill="none" stroke="currentColor" stroke-width="3"
+                            d="M 500 650 L 500 800 L 150 800 L 150 70 L 200 70 L 250 70" />
+                        <path id="path17" fill="none" stroke="currentColor" stroke-width="3"
+                            d="M 120 20 L 120 60 L 250 60 " />
+                        <rect id="MUX1" x="240" y="220" width="40" height="40"
```
Comment (raj-vlabs) [inline experiment/simulation/demo-parallel-register.html]: Please use lowercase identifiers

File: experiment/simulation/js/animation-parallel-register.js [hunk 14]
```diff
@@ -0,0 +1,807 @@
...
+const I3 = document.getElementById("SERIALINPUT3");
+const I4 = document.getElementById("SERIALINPUT4");
+const CLOCK = document.getElementById("CLOCK");
+const O1 = document.getElementById("SERIALOUTPUT1");
+const O2 = document.getElementById("SERIALOUTPUT2");
+const O3 = document.getElementById("SERIALOUTPUT3");
+const O4 = document.getElementById("SERIALOUTPUT4");
+const SELECT = document.getElementById("COMMONSELECT");
+const BUTTON = document.getElementById("play/pause");
+const OBSERV = document.getElementById("Observations");
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:78]: Please do not use uppercase identifiers unless they represent a constant value.

File: experiment/simulation/js/animation-parallel-register.js [hunk 15]
```diff
@@ -0,0 +1,807 @@
...
+});
+let serialDot2 = document.createElementNS(svgns, "circle");
+gsap.set(serialDot2, {
+    attr: { cx: 50, cy: 250, r: 15, fill: "#FF0000" }
+});
+let serialDot3 = document.createElementNS(svgns, "circle");
+gsap.set(serialDot3, {
+    attr: { cx: 50, cy: 450, r: 15, fill: "#FF0000" }
+});
+let serialDot4 = document.createElementNS(svgns, "circle");
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:96]: Too many identifiers defined with let, most of them look const

File: experiment/simulation/js/animation-parallel-register.js [hunk 16]
```diff
@@ -0,0 +1,807 @@
...
+    attr: { cx: 50, cy: 650, r: 15, fill: "#FF0000" }
+});
+
+svg.appendChild(serialDot1);
+svg.appendChild(serialDot3);
+svg.appendChild(serialDot4);
+svg.appendChild(clockDot);
+svg.appendChild(serialDot2);
+
+function myFunction() {
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:107]: Please change this name myFunction to be meaningful

File: experiment/simulation/js/animation-parallel-register.js [hunk 17]
```diff
@@ -0,0 +1,807 @@
...
+    OBSERV.innerHTML = "";
+}
+function selectTextDisappear()
+{
+    TweenLite.to(textSelect, 0, { autoAlpha: 0 });
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:112]: TweenLite has been deprecated in GSAP3, please remove all references to it

File: experiment/simulation/js/animation-parallel-register.js [hunk 18]
```diff
@@ -0,0 +1,807 @@
...
+    serialDisappear1();
+    serialDisappear2();
+    serialDisappear3();
+    serialDisappear4();
+    serialDotDisappear();
+    selectDisappear();
+    clockDisappear();
+    clockDotDisappear();
+    outputDisappear();
+    gsap.set(I1, {
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:227]: In place of calling the set function umpteen times , can we just create an array of all these elements and use a forEach or for . of .. loop to set all of them to get much readable code?

File: experiment/simulation/js/animation-parallel-register.js [hunk 19]
```diff
@@ -0,0 +1,807 @@
...
+
+            fill: "#eeeb22"
+        });
+        setter(textI3.textContent, serialDot3);
+        free();
+        serialVisible3();
+        errno();
+
+
+
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:425]: remove unnecessary empty lines

File: experiment/simulation/js/animation-parallel-register.js [hunk 20]
```diff
@@ -0,0 +1,807 @@
...
+        OBSERV.innerHTML = "I3 is set to 0";
+    }
+    else if (textI3.textContent != 1 && tl.progress() == 0) {
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:428]: == and != is not what we recommend for comparison. Please go through the coding guidelines

File: experiment/simulation/js/animation-parallel-register.js [hunk 21]
```diff
@@ -0,0 +1,807 @@
...
+    setter(textO4.textContent, O4);
+}
+
+function errno() {
+
+}
+function batado() {
+    OBSERV.innerHTML = "Simulation has finished. Press Restart to start again"
+}
+function setter(a, b) {
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:577]: What am I expected to understand about this function from it name?

File: experiment/simulation/js/animation-parallel-register.js [hunk 22]
```diff
@@ -0,0 +1,807 @@
...
+    if (a == 1) {
+        unset(b);
+
+    }
+    else if (a == 0) {
+        set(b);
+    }
+}
+outputDisappear();
+var tl = gsap.timeline({ repeat: 0, repeatDelay: 0 });
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:587]: Why are we using var?

File: experiment/simulation/js/animation-parallel-register.js [hunk 23]
```diff
@@ -0,0 +1,807 @@
...
+function fourXspeed() {
+    if (textClock.textContent == 1 && textI1.textContent != 2 && textI2.textContent != 2 && textI3.textContent != 2 && textI4.textContent != 2 && tl.progress() != 1) {
+        tl.resume();
+        tl.timeScale(4);
+        OBSERV.innerHTML = "4x speed";
+        decide = 1;
+        BUTTON.innerHTML = "Halt";
+    }
+}
+function doubleSpeed() {
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js:598]: Do we need 2 separate functions with almost identical code for 2x and 4x speeds?

File: experiment/simulation/js/animation-parallel-register.js [hunk 24]
```diff
@@ -0,0 +1,807 @@
...
+    if (textClock.textContent == 1 && textI1.textContent != 2 && textI2.textContent != 2 && textI3.textContent != 2 && textI4.textContent != 2 && tl.progress() != 1) {
+        tl.resume();
+        tl.timeScale(2);
+        OBSERV.innerHTML = "2x speed";
+        decide = 1;
+        BUTTON.innerHTML = "Halt";
+    }
+}
+function SetSpeed(speed) {
```
Comment (raj-vlabs) [inline experiment/simulation/js/animation-parallel-register.js]: Function names are recommended to start with lowercase.

**Review by raj-vlabs — CHANGES_REQUESTED** [review]
The code needs major overhaul. Overall its very disappointing to see this quality of code. Basic coding guidelines have not been followed, naming convention is arbitrary, recommended comparison operators have not been used, no thought has been given to make the code readable and compact, deprecated library features have been used, deprecated html elements have been used etc etc.
I have not reviewed all the code and I don't see any point of doing so. Please ensure that the basics are followed for the whole code and resubmit for review.

File: experiment/simulation/demo.html [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+            <button class="v-button" id="play-or-pause" onclick="simulationStatus()">Start</button>
+            <button class="v-button" id="restart" onclick="restartCircuit()">Reset</button>
+        </div>
+    </div>
+    <div class="columns is-centered is-flex-direction-row is-flex-wrap-wrap is-align-content-start">
+        <div class="column is-9-desktop is-12-tablet is-12-mobile">
+            <div id="circuit-board">
+                <div id="working-area">
+
+                    <svg class="svg" xmlns="http://www.w3.org/2000/svg" width="100%" height="140%"
```
Comment (raj-vlabs) [inline experiment/simulation/demo.html]: Please change all the links to https.

File: experiment/simulation/js/anim.js [hunk 26]
```diff
@@ -0,0 +1,484 @@
...
+const circuitBoard = document.getElementById("circuit-board");
+const sidePanels = document.getElementsByClassName("v-datalist-container");
+// Distance of working area from top
+const circuitBoardTop = circuitBoard.offsetTop;
+// Full height of window
+const windowHeight = window.innerHeight;
+const width = window.innerWidth;
+
+const svg = document.querySelector(".svg");
+const svgns = "http://www.w3.org/2000/svg";
```
Comment (raj-vlabs) [inline experiment/simulation/js/anim.js:28]: This link will probably give errors if you change it to https, in that case we can keep it to http till we are able to fix the errors.

**Review by raj-vlabs — CHANGES_REQUESTED** [review]
Please change all the links to https except for the on used as svgns.

I generally like the code quality. The readability has improved significantly.

File: experiment/simulation/demo.html (hunk 25 above)
Comment (shreyash-x) [reply experiment/simulation/demo.html]: Was just able to do it in the HTML files, gave errors everywhere else.

**Review by raj-vlabs — APPROVED** [review]

**Pull request merged**